src/basic/fd-util.c

```diff
@@ -433,11 +433,9 @@ bool fdname_is_valid(const char *s) {
 }
 
 int fd_get_path(int fd, char **ret) {
-        char procfs_path[STRLEN("/proc/self/fd/") + DECIMAL_STR_MAX(int)];
         int r;
 
-        xsprintf(procfs_path, "/proc/self/fd/%i", fd);
-        r = readlink_malloc(procfs_path, ret);
+        r = readlink_malloc(FORMAT_PROC_FD_PATH(fd), ret);
```

Comment:
I think it is better to add `assert(fd >= 0);`

Reply:
Or instead, maybe move the assert *into* `FORMAT_PROC_FD_PATH`? I didn't go over all the uses, but it seems we always expect to pass a valid fd to the macro, and usually immediately use the result in a further call that'd crash if it were to get a NULL.